sky/provision/do/instance.py

```diff
@@ -0,0 +1,299 @@
+            body={'type': 'shutdown'},
+        )
+
+    # Wait for instances to stop
+    for _ in range(MAX_POLLS_FOR_UP_OR_STOP):
+        all_instances = utils.filter_instances(cluster_name_on_cloud, ['off'])
+        if len(all_instances) >= num_instances:
+            break
+        time.sleep(constants.POLL_INTERVAL)
+    else:
+        raise RuntimeError(f'Maximum number of polls: '
+                           f'{MAX_POLLS_FOR_UP_OR_STOP} reached. '
+                           f'Instance {all_instances} is still not in '
+                           'STOPPED status.')
+
+
+def terminate_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    """See sky/provision/__init__.py"""
+    del provider_config  # unused
+    instances = utils.filter_instances(cluster_name_on_cloud,
+                                       status_filters=None)
+    for instance_name, instance_meta in instances.items():
+        logger.debug(f'Terminating instance {instance_name}')
+        if worker_only and instance_name.endswith('-head'):
+            continue
+        try:
+            utils.client().droplets.destroy_with_associated_resources_dangerous(
+                droplet_id=instance_meta['id'], x_dangerous=True)
+        except azure.exceptions().HttpResponseError as err:
+            logger.warning('Error: {0} {1}: {2}'.format(err.status_code,
+                                                        err.reason,
+                                                        err.error.message))
+
+    for _ in range(MAX_POLLS_FOR_UP_OR_STOP):
+        instances = utils.filter_instances(cluster_name_on_cloud,
+                                           status_filters=None)
+        if len(instances) == 0 or len(instances) <= 1 and worker_only:
+            break
+        time.sleep(constants.POLL_INTERVAL)
+    else:
+        msg = ('Failed to delete all instances')
+        logger.warning(msg)
+        raise RuntimeError(msg)
+
+
+def get_cluster_info(
+    region: str,
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> common.ClusterInfo:
+    del region  # unused
+    running_instances = utils.filter_instances(cluster_name_on_cloud,
+                                               ['active'])
+    instances: Dict[str, List[common.InstanceInfo]] = {}
+    head_instance: Optional[Dict[str, Any]] = None
+    for instance_name, instance_meta in running_instances.items():
+        if instance_name.endswith('-head'):
+            head_instance = instance_meta
+        for net in instance_meta['networks']['v4']:
+            if net['type'] == 'public':
+                instance_ip = net['ip_address']
+                break
+        instances[instance_name] = [
+            common.InstanceInfo(
+                instance_id=instance_meta['name'],
+                internal_ip=instance_ip,
+                external_ip=instance_ip,
```

Comment:
the short answer is multinode breaks if we use the internal ip here (I already tried this).

the long answer is the public ip address is exposed as the first ip address on every node. The ray executor compares the `internal_ip` against the one's ray discovers. However, ray will discover the first ip address it sees on each node, which is `eth0` ip address which is the public one, the executor will see that there is a mismatch in the internal ips given by skypilot and the ones ray discovers and execution will fail.

```
(base) root@sky-eb1b-gcpuser-323b-0a28-head:~# ip -br addr
lo               UNKNOWN        127.0.0.1/8 ::1/128 
eth0             UP             167.172.255.36/20 10.17.0.5/16 fe80::a017:c0ff:fe07:6e1f/64 
eth1             UP             10.108.0.2/20 fe80::a0db:f7ff:fea5:82e3/64 
docker0          DOWN           172.17.0.1/16
```

Reply:
So ray will always treat eth0 as the private ip address? is there any method to customize it?

Reply:
i want to say no. we use `ray.util.get_node_ip_address` [here](https://github.com/skypilot-org/skypilot/blob/057bc4b44755ac1e9dadc680e022c369e8ddff52/sky/backends/cloud_vm_ray_backend.py#L499), and it doesn't really expose any arguments that would let us bypass the first detected ip interface. 
https://github.com/ray-project/ray/blob/f44890f90284c41c4d26a2d1e3b36b3ce4a8f4ff/python/ray/_private/services.py#L677

I also tried taking down the `eth0` interface, but this will make the instance unreachable from the public internet

Reply:
I see. looks good to me. cc @Michaelvll for a look here